src/tactic/norm_cast.lean

```diff
@@ -25,225 +25,314 @@ try_for 1000 (mk_instance e)
-        return ((), new_e, pr)
-    )
-| _ := failed
-end
+private meta def prove : tactic unit :=
+assumption
 
 /-
-simpa is used to discharge proofs
+TODO: norm_cast takes a list of expressions to use as lemmas for the discharger
 -/
-private meta def prove : tactic unit :=
-tactic.interactive.simpa none ff [] [] none
 
+/-
+This is an auxiliary function used in step 2.
+It tries to rewrite an expression using the elim_cast and move_cast lemmas.
+On failure, it calls the heuristic.
+-/
 private meta def post (s : simp_lemmas) (_ : unit) (e : expr) : tactic (unit × expr × expr) :=
-do
-    r ← mcond (is_prop e) (return `iff) (return `eq),
-    (new_e, pr) ← s.rewrite e prove r,
-    pr ← match r with
-    |`iff := mk_app `propext [pr]
-    | _   := return pr
-    end,
-    return ((), new_e, pr)
+( do
+  r ← mcond (is_prop e) (return `iff) (return `eq),
+  (new_e, pr) ← s.rewrite e prove r,
+  pr ← match r with
+  | `iff := mk_app `propext [pr]
+  | _    := return pr
+  end,
+  return ((), new_e, pr)
+) <|> heur () e
+
+/-
+The following auxiliary functions are used to handle numerals.
+-/
 
+-- prove ↑n = n where n is a numeral
+private meta def aux_num_prove_eq (a b : expr) : tactic expr :=
+do
+  s1 ← simp_lemmas.mk_default,
+  s2 ← squash_cast_attr.get_cache,
+  let s := simp_lemmas.join s1 s2,
+  h ← to_expr ``(%%a = %%b),
+  (_, pr) ← simplify s [] h,
+  to_expr ``(eq.mpr %%pr trivial)
+
+-- if possible, rewrite (n : α) to ((n : ℕ) : α) where n is a numeral and α ≠ ℕ
+private meta def aux_num_1 (_ : unit) (e : expr) : tactic (unit × expr × expr) :=
+do
+  α ← infer_type e,
+  success_if_fail $ is_def_eq α `(ℕ),
+  n ← e.to_num,
+  h1 ← mk_app `has_lift_t [`(ℕ), α] >>= mk_instance',
+  new_e ← to_expr $ pexpr_of_num `(ℕ) n,
+  new_e ← to_expr ``(@coe ℕ %%α %%h1 %%new_e),
+  pr ← aux_num_prove_eq e new_e,
+  return ((), new_e, pr)
+
+-- if possible, rewrite (↑n : α) to (n : α) where n is a numeral
+private meta def aux_num_2 (_ : unit) (e : expr) : tactic (unit × expr × expr) :=
+do
+  `(@coe ℕ %%α %%h1 %%e') ← return e,
+  n ← e'.to_num,
+  new_e ← to_expr $ pexpr_of_num α n,
+  pr ← aux_num_prove_eq e new_e,
+  return ((), new_e, pr)
+
+private meta def simplify_top_down' {α} (a : α) (pre : α → expr → tactic (α × expr × expr)) (e : expr) (cfg : simp_config := {}) : tactic (α × expr × expr) :=
+ext_simplify_core a cfg simp_lemmas.mk (λ _, failed)
+  (λ a _ _ _ e, do (new_a, new_e, pr) ← pre a e, guard (¬ new_e =ₐ e), return (new_a, new_e, some pr, ff))
+  (λ _ _ _ _ _, failed)
+  `eq e
 
 /-
-Core function
+Core function.
 -/
 meta def derive (e : expr) : tactic (expr × expr) :=
 do
-    s ← get_cache,
-    e ← instantiate_mvars e,
-    let cfg : simp_config := {fail_if_unchanged := ff},
+  s ← get_cache,
+  e ← instantiate_mvars e,
+  let cfg : simp_config := { fail_if_unchanged := ff },
+  let e0 := e,
+
+  -- step 1: pre-processing of numerals
+  ((), e1, pr1) ← simplify_top_down' () aux_num_1 e0 cfg,
+  h1 ← to_expr ``(%%e0 = %%e1),
```

Comment:
`h1` isn't used below, nor is `h4`.

Reply:
must have been used for debug at some point, to check the type of pr1 and pr4, I'll remove them